db-schema.sql

```diff
@@ -292,6 +242,60 @@ CREATE TABLE ApplicationState (
 
 );
 
+-- Represents relationship from GitOpsDeployment CR in the namespace, to an Application table row 
+-- This means: if we see a change in a GitOpsDeployment CR, we can easily find the corresponding database entry
+-- Also: if we see a change to an Argo CD Application, we can easily find the corresponding GitOpsDeployment CR
+CREATE TABLE DeploymentToApplicationMapping (
+	
+	-- uid of our gitops deployment CR within the K8s namespace (or KCP control plane)
+	deploymenttoapplicationmapping_uid_id VARCHAR(48) UNIQUE NOT NULL PRIMARY KEY,
+
+	-- Foreign key to: Application.application_id
+	application_id VARCHAR ( 48 ) NOT NULL UNIQUE,
+	CONSTRAINT fk_app_id FOREIGN KEY (application_id) REFERENCES Application(application_id) ON DELETE CASCADE ON UPDATE CASCADE,
+
+	seq_id serial
+
+);
+
+-- Represents a generic relationship between Kubernetes CR <-> Database table
+-- Useful for tracking the lifecycle between the two.
+--
+-- Example: when the cluster agent sees an Argo CD Application CR change within a namespace, it needs a way
+-- to know which GitOpsEngineInstance database entries corresponds to the Argo CD namespace.
+-- For this we would use:
+-- - kubernetes_resource_type: Namespace
+-- - kubernetes_resource_uid: (uid of namespace)
+-- - db_relation_type: GitOpsEngineInstance
+-- - db_relation_key: (primary key of gitops engine instance)
+--
+-- Later, we can query this table to determine 'argo cd instance namespace' <=> 'GitopsEngineInstance database row'
+--
+-- This is also useful for tracking the lifecycle between CRs <-> database table.
+CREATE TABLE KubernetesToDBResourceMapping  (
+
+	kubernetes_resource_type VARCHAR(64) NOT NULL,
+
+	kubernetes_resource_uid  VARCHAR(64) NOT NULL,
+
+	db_relation_type  VARCHAR(64) NOT NULL,
+
+	db_relation_key  VARCHAR(64) NOT NULL,
+
+	seq_id serial,
+
+	PRIMARY KEY(kubernetes_resource_type, kubernetes_resource_uid, db_relation_type, db_relation_key)
+);
+
+-- TODO: Add index for
+-- - kubernetes_resource_type VARCHAR(64) NOT NULL,
+-- - kubernetes_resource_uid  VARCHAR(64) NOT NULL,
+-- - db_relation_type  VARCHAR(64) NOT NULL,
+
+CREATE INDEX idx_resource_type ON KubernetesToDBResourceMapping(kubernetes_resource_type);
```

Comment:
Whoops, I should have been more descriptive in my TODO! This should be a single index containing all three fields mentioned in the TODO. This allows us to do quick lookups  of (k8s resource type, k8s resource uid, db_relation_type) -> db_relation_uid, which I have found I am doing a fair bit of in the code at the moment.

Reply:
Makes sense! :hugs: